/*
 * SPDX-License-Identifier: MIT
 * SPDX-FileCopyrightText: Copyright (c) 2024 OpenStickCommunity (gp2040-ce.info)
 */

#ifndef MAVERCADE_CONFIG_H_
#define MAVERCADE_CONFIG_H_

#include "enums.pb.h"
#include "class/hid/hid.h"

#define BOARD_CONFIG_LABEL "Mavercade Rev2"

// Main pin mapping Configuration
//                          // GP2040 | Xinput | Switch  | PS3/4/5  | Dinput | Arcade |
#define GPIO_PIN_11 GpioAction::BUTTON_PRESS_UP     // UP     | UP     | UP      | UP       | UP     | UP     |
#define GPIO_PIN_20 GpioAction::BUTTON_PRESS_UP     // UP     | UP     | UP      | UP       | UP     | UP     |
#define GPIO_PIN_08 GpioAction::BUTTON_PRESS_DOWN   // DOWN   | DOWN   | DOWN    | DOWN     | DOWN   | DOWN   |
#define GPIO_PIN_10 GpioAction::BUTTON_PRESS_RIGHT  // RIGHT  | RIGHT  | RIGHT   | RIGHT    | RIGHT  | RIGHT  |
#define GPIO_PIN_07 GpioAction::BUTTON_PRESS_LEFT   // LEFT   | LEFT   | LEFT    | LEFT     | LEFT   | LEFT   |
#define GPIO_PIN_12 GpioAction::BUTTON_PRESS_B1     // B1     | A      | B       | Cross    | 2      | K1     |
#define GPIO_PIN_17 GpioAction::BUTTON_PRESS_B2     // B2     | B      | A       | Circle   | 3      | K2     |
#define GPIO_PIN_18 GpioAction::BUTTON_PRESS_R2     // R2     | RT     | ZR      | R2       | 8      | K3     |
#define GPIO_PIN_09 GpioAction::BUTTON_PRESS_L2     // L2     | LT     | ZL      | L2       | 7      | K4     |
#define GPIO_PIN_16 GpioAction::BUTTON_PRESS_B3     // B3     | X      | Y       | Square   | 1      | P1     |
#define GPIO_PIN_14 GpioAction::BUTTON_PRESS_B4     // B4     | Y      | X       | Triangle | 4      | P2     |
#define GPIO_PIN_15 GpioAction::BUTTON_PRESS_R1     // R1     | RB     | R       | R1       | 6      | P3     |
#define GPIO_PIN_19 GpioAction::BUTTON_PRESS_L1     // L1     | LB     | L       | L1       | 5      | P4     |
#define GPIO_PIN_23 GpioAction::BUTTON_PRESS_S1     // S1     | Back   | Minus   | Select   | 9      | Coin   |
#define GPIO_PIN_24 GpioAction::BUTTON_PRESS_S2     // S2     | Start  | Plus    | Start    | 10     | Start  |
#define GPIO_PIN_06 GpioAction::BUTTON_PRESS_L3     // L3     | LS     | LS      | L3       | 11     | LS     |
#define GPIO_PIN_13 GpioAction::BUTTON_PRESS_R3     // R3     | RS     | RS      | R3       | 12     | RS     |
#define GPIO_PIN_22 GpioAction::BUTTON_PRESS_A1     // A1     | Guide  | Home    | PS       | 13     | ~      |
#define GPIO_PIN_21 GpioAction::BUTTON_PRESS_A2     // A2     | ~      | Capture | ~        | 14     | ~      |

// Setting GPIO pins to assigned by add-on
//
#define GPIO_PIN_01 GpioAction::ASSIGNED_TO_ADDON
#define GPIO_PIN_02 GpioAction::ASSIGNED_TO_ADDON
#define GPIO_PIN_03 GpioAction::ASSIGNED_TO_ADDON
#define GPIO_PIN_04 GpioAction::ASSIGNED_TO_ADDON
#define GPIO_PIN_05 GpioAction::ASSIGNED_TO_ADDON
#define GPIO_PIN_29 GpioAction::ASSIGNED_TO_ADDON

// Keyboard Mapping Configuration
//                                            // GP2040 | Xinput | Switch  | PS3/4/5  | Dinput | Arcade |
#define KEY_DPAD_UP     HID_KEY_ARROW_UP      // UP     | UP     | UP      | UP       | UP     | UP     |
#define KEY_DPAD_DOWN   HID_KEY_ARROW_DOWN    // DOWN   | DOWN   | DOWN    | DOWN     | DOWN   | DOWN   |
#define KEY_DPAD_RIGHT  HID_KEY_ARROW_RIGHT   // RIGHT  | RIGHT  | RIGHT   | RIGHT    | RIGHT  | RIGHT  |
#define KEY_DPAD_LEFT   HID_KEY_ARROW_LEFT    // LEFT   | LEFT   | LEFT    | LEFT     | LEFT   | LEFT   |
#define KEY_BUTTON_B1   HID_KEY_SHIFT_LEFT    // B1     | A      | B       | Cross    | 2      | K1     |
#define KEY_BUTTON_B2   HID_KEY_Z             // B2     | B      | A       | Circle   | 3      | K2     |
#define KEY_BUTTON_R2   HID_KEY_X             // R2     | RT     | ZR      | R2       | 8      | K3     |
#define KEY_BUTTON_L2   HID_KEY_V             // L2     | LT     | ZL      | L2       | 7      | K4     |
#define KEY_BUTTON_B3   HID_KEY_CONTROL_LEFT  // B3     | X      | Y       | Square   | 1      | P1     |
#define KEY_BUTTON_B4   HID_KEY_ALT_LEFT      // B4     | Y      | X       | Triangle | 4      | P2     |
#define KEY_BUTTON_R1   HID_KEY_SPACE         // R1     | RB     | R       | R1       | 6      | P3     |
#define KEY_BUTTON_L1   HID_KEY_C             // L1     | LB     | L       | L1       | 5      | P4     |
#define KEY_BUTTON_S1   HID_KEY_5             // S1     | Back   | Minus   | Select   | 9      | Coin   |
#define KEY_BUTTON_S2   HID_KEY_1             // S2     | Start  | Plus    | Start    | 10     | Start  |
#define KEY_BUTTON_L3   HID_KEY_EQUAL         // L3     | LS     | LS      | L3       | 11     | LS     |
#define KEY_BUTTON_R3   HID_KEY_MINUS         // R3     | RS     | RS      | R3       | 12     | RS     |
#define KEY_BUTTON_A1   HID_KEY_9             // A1     | Guide  | Home    | PS       | 13     | ~      |
#define KEY_BUTTON_A2   HID_KEY_F2            // A2     | ~      | Capture | ~        | 14     | ~      |
#define KEY_BUTTON_FN   -1                    // Hotkey Function                                        |

#define USB_PERIPHERAL_ENABLED 1
#define USB_PERIPHERAL_PIN_DPLUS 1

#define DEFAULT_INPUT_MODE_R1 INPUT_MODE_XBONE
#define DEFAULT_INPUT_MODE_B4 INPUT_MODE_PS5
#define DEFAULT_PS5AUTHENTICATION_TYPE INPUT_MODE_AUTH_TYPE_USB

#define BOARD_LEDS_PIN 3

#define LED_BRIGHTNESS_MAXIMUM 50
#define LEDS_PER_PIXEL 1
#define LEDS_BASE_ANIMATION_INDEX 1
#define LEDS_BUTTON_A2   0
#define LEDS_BUTTON_L3   1
#define LEDS_DPAD_LEFT   2
#define LEDS_DPAD_DOWN   3
#define LEDS_DPAD_RIGHT  4
#define LEDS_BUTTON_B3   5
#define LEDS_BUTTON_B4   6
#define LEDS_BUTTON_R1   7
#define LEDS_BUTTON_L1   8
#define LEDS_BUTTON_L2   9
#define LEDS_BUTTON_R2   10
#define LEDS_BUTTON_B2   11
#define LEDS_BUTTON_B1   12
#define LEDS_BUTTON_R3   13
#define LEDS_DPAD_UP     14

#define HAS_I2C_DISPLAY 1
#define I2C0_ENABLED 1
#define I2C0_PIN_SDA 4
#define I2C0_PIN_SCL 5
#define DISPLAY_I2C_BLOCK i2c0

#define REVERSE_UP_DEFAULT 1
#define REVERSE_DOWN_DEFAULT 1
#define REVERSE_LEFT_DEFAULT 1
#define REVERSE_RIGHT_DEFAULT 1

#define SPLASH_MODE SPLASH_MODE_STATIC
#define SPLASH_DURATION 3000

#define BUTTON_LAYOUT BUTTON_LAYOUT_BOARD_DEFINED_A
#define BUTTON_LAYOUT_RIGHT BUTTON_LAYOUT_BOARD_DEFINED_B

// Keebbrawler 04 Updated
#define DEFAULT_BOARD_LAYOUT_A {\
  {GP_ELEMENT_PIN_BUTTON, {5,  17, 12, 33, 1, 1, 6,      GP_SHAPE_SQUARE}},\
  {GP_ELEMENT_PIN_BUTTON, {19, 23,  4,  4, 1, 1, 7,      GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {29, 23,  4,  4, 1, 1, 8,      GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {33, 14,  4,  4, 1, 1, 6,      GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {40, 25,  4,  4, 1, 1, 10,     GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {25, 50, 40, 43, 1, 1, 19,     GP_SHAPE_SQUARE, 135}},\
  {GP_ELEMENT_PIN_BUTTON, {38, 50, 53, 43, 1, 1, 11,     GP_SHAPE_SQUARE, 135}}\
}

#define DEFAULT_BOARD_LAYOUT_B {\
  {GP_ELEMENT_PIN_BUTTON, {88,  21,   4,  4, 1, 1, 16,   GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {99,  19,   4,  4, 1, 1, 14,   GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {109, 19,   4,  4, 1, 1, 15,   GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {87,  32,   4,  4, 1, 1, 12,   GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {98,  30,   4,  4, 1, 1, 17,   GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {108, 30,   4,  4, 1, 1, 18,   GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {82,  50,  97, 43, 1, 1, 9,    GP_SHAPE_SQUARE, 45}},\
  {GP_ELEMENT_PIN_BUTTON, {69,  50,  84, 43, 1, 1, 20,   GP_SHAPE_SQUARE, 45}},\
  {GP_ELEMENT_PIN_BUTTON, {123, 17, 116, 33, 1, 1, 13,   GP_SHAPE_SQUARE}}\
}

// Keebbrawler 03
#define DEFAULT_BOARD_LAYOUT_A_ALT0 {\
  {GP_ELEMENT_PIN_BUTTON, {5,  17, 12, 33, 1, 1, 6,      GP_SHAPE_SQUARE}},\
  {GP_ELEMENT_PIN_BUTTON, {19, 23,  4,  4, 1, 1, 7,      GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {29, 23,  4,  4, 1, 1, 8,      GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {40, 25,  4,  4, 1, 1, 10,     GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {25, 50, 42, 43, 1, 1, 19,     GP_SHAPE_SQUARE, 135}},\
  {GP_ELEMENT_PIN_BUTTON, {45, 40, 79, 50, 1, 1, 11,     GP_SHAPE_SQUARE}}\
}
#define DEFAULT_BOARD_LAYOUT_B_ALT0 {\
  {GP_ELEMENT_PIN_BUTTON, {88,  21,   4,  4, 1, 1, 16,   GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {99,  19,   4,  4, 1, 1, 14,   GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {109, 19,   4,  4, 1, 1, 15,   GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {87,  32,   4,  4, 1, 1, 12,   GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {98,  30,   4,  4, 1, 1, 17,   GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {108, 30,   4,  4, 1, 1, 18,   GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {82,  50,  99, 43, 1, 1, 9,    GP_SHAPE_SQUARE, 45}},\
  {GP_ELEMENT_PIN_BUTTON, {123, 17, 116, 33, 1, 1, 13,   GP_SHAPE_SQUARE}}\
}

// Keebfighter S3 Updated
#define DEFAULT_BOARD_LAYOUT_A_ALT1 {\
  {GP_ELEMENT_PIN_BUTTON, {19, 23, 4, 4, 1, 1, 7,        GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {29, 23, 4, 4, 1, 1, 8,        GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {33, 14, 4, 4, 1, 1, 20,       GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {40, 25, 4, 4, 1, 1, 10,       GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {45, 46, 7, 7, 1, 1, 11,       GP_SHAPE_ELLIPSE}}\
}
#define DEFAULT_BOARD_LAYOUT_B_ALT1 {\
  {GP_ELEMENT_PIN_BUTTON, {88,  21, 4, 4, 1, 1, 16,      GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {99,  19, 4, 4, 1, 1, 14,      GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {109, 20, 4, 4, 1, 1, 15,      GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {119, 21, 4, 4, 1, 1, 19,      GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {87,  32, 4, 4, 1, 1, 12,      GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {98,  30, 4, 4, 1, 1, 17,      GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {108, 31, 4, 4, 1, 1, 18,      GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {118, 32, 4, 4, 1, 1, 9,       GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {95,  41, 4, 4, 1, 1, 6,       GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {81,  46, 7, 7, 1, 1, 13,      GP_SHAPE_ELLIPSE}}\
}

// Keebfighter S3
#define DEFAULT_BOARD_LAYOUT_A_ALT2 {\
  {GP_ELEMENT_PIN_BUTTON, {19, 23, 4, 4, 1, 1, 7,        GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {29, 23, 4, 4, 1, 1, 8,        GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {33, 14, 4, 4, 1, 1, 20,       GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {40, 25, 4, 4, 1, 1, 10,       GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {9,  28, 4, 4, 1, 1, 6,        GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {45, 46, 7, 7, 1, 1, 11,       GP_SHAPE_ELLIPSE}}\
}

#define DEFAULT_BOARD_LAYOUT_B_ALT2 {\
  {GP_ELEMENT_PIN_BUTTON, {88,  21, 4, 4, 1, 1, 16,      GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {99,  19, 4, 4, 1, 1, 14,      GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {109, 20, 4, 4, 1, 1, 15,      GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {119, 21, 4, 4, 1, 1, 19,      GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {87,  32, 4, 4, 1, 1, 12,      GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {98,  30, 4, 4, 1, 1, 17,      GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {108, 31, 4, 4, 1, 1, 18,      GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {118, 32, 4, 4, 1, 1, 9,       GP_SHAPE_ELLIPSE}},\
  {GP_ELEMENT_PIN_BUTTON, {81,  46, 7, 7, 1, 1, 13,      GP_SHAPE_ELLIPSE}}\
}

// Keebfighter 02
#define DEFAULT_BOARD_LAYOUT_A_ALT2 {\
  {GP_ELEMENT_PIN_BUTTON, {23, 25, 16, 18, 1, 1, 26,     GP_SHAPE_SQUARE, 22}},\
  {GP_ELEMENT_PIN_BUTTON, {32, 27, 25, 20, 1, 1, 6,      GP_SHAPE_SQUARE, 22}},\
  {GP_ELEMENT_PIN_BUTTON, {40, 31, 33, 24, 1, 1, 25,     GP_SHAPE_SQUARE, 22}},\
  {GP_ELEMENT_PIN_BUTTON, {47, 37, 40, 30, 1, 1, 13,     GP_SHAPE_SQUARE, 22}},\
  {GP_ELEMENT_PIN_BUTTON, {19, 33, 12, 26, 1, 1, 27,     GP_SHAPE_SQUARE, 22}},\
  {GP_ELEMENT_PIN_BUTTON, {28, 35, 21, 28, 1, 1, 7,      GP_SHAPE_SQUARE, 22}},\
  {GP_ELEMENT_PIN_BUTTON, {36, 39, 29, 32, 1, 1, 8,      GP_SHAPE_SQUARE, 22}},\
  {GP_ELEMENT_PIN_BUTTON, {43, 45, 36, 38, 1, 1, 10,     GP_SHAPE_SQUARE, 22}},\
  {GP_ELEMENT_PIN_BUTTON, {40, 54, 55, 47, 1, 1, 11,     GP_SHAPE_SQUARE, 135}}\
}

#define DEFAULT_BOARD_LAYOUT_B_ALT2 {\
  {GP_ELEMENT_PIN_BUTTON, {88,  30,  81, 37, 1, 1, 16,   GP_SHAPE_SQUARE, 158}},\
  {GP_ELEMENT_PIN_BUTTON, {95,  24,  88, 31, 1, 1, 14,   GP_SHAPE_SQUARE, 158}},\
  {GP_ELEMENT_PIN_BUTTON, {103, 20,  96, 27, 1, 1, 15,   GP_SHAPE_SQUARE, 158}},\
  {GP_ELEMENT_PIN_BUTTON, {112, 18, 105, 25, 1, 1, 19,   GP_SHAPE_SQUARE, 158}},\
  {GP_ELEMENT_PIN_BUTTON, {92,  38,  85, 45, 1, 1, 12,   GP_SHAPE_SQUARE, 158}},\
  {GP_ELEMENT_PIN_BUTTON, {99,  32,  92, 39, 1, 1, 17,   GP_SHAPE_SQUARE, 158}},\
  {GP_ELEMENT_PIN_BUTTON, {107, 28, 100, 35, 1, 1, 18,   GP_SHAPE_SQUARE, 158}},\
  {GP_ELEMENT_PIN_BUTTON, {116, 26, 109, 33, 1, 1, 9,    GP_SHAPE_SQUARE, 158}},\
  {GP_ELEMENT_PIN_BUTTON, {73,  54,  88, 47, 1, 1, 20,   GP_SHAPE_SQUARE, 45}}\
}

#define FOCUS_MODE_ENABLED 1
#define FOCUS_MODE_BUTTON_MASK GAMEPAD_MASK_S1 | GAMEPAD_MASK_S2 | GAMEPAD_MASK_A1 | GAMEPAD_MASK_A2 
#define FOCUS_MODE_PIN 29
#define FOCUS_MODE_BUTTON_LOCK_ENABLED 1

#define DEFAULT_SPLASH \
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x00,0x10,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x00,0x10,0x00,0x00,0x00,0x00,0x40,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x70,0x00,0x00,0x30,0x00,0x00,0x00,0x00,0xE0,0x00,0x00,0x00,0x00,0x00, \
  0x00,0xE1,0xF0,0x00,0x00,0x70,0x00,0x00,0x00,0x00,0xE0,0x00,0x00,0x00,0x00,0x00, \
  0x03,0xE1,0xF0,0x3F,0xE0,0x61,0xFF,0x8C,0x00,0x06,0xF0,0x3E,0x1F,0x00,0x3F,0xE0, \
  0x03,0xE3,0xF0,0x7F,0xE0,0xDF,0xFF,0xEF,0xF8,0x0F,0xF0,0x7C,0xFF,0xC7,0xFF,0xFC, \
  0x03,0xE3,0xF0,0xFF,0xE1,0xDF,0xFF,0xDF,0xFC,0x1F,0xF0,0xFD,0xFF,0xE7,0xFF,0xF8, \
  0x03,0xC7,0xE1,0xFF,0xE1,0xDF,0xF8,0x3F,0xFE,0x3F,0xE0,0xFF,0xF1,0xF3,0xFF,0x00, \
  0x07,0xC7,0xE1,0xFB,0xE3,0xC7,0x00,0x1E,0x0F,0x3D,0xE1,0xFC,0x70,0xF8,0xE0,0x00, \
  0x07,0xCF,0xC3,0xF9,0xE3,0xC7,0x00,0x1E,0x1F,0x7D,0x83,0xF8,0x70,0x79,0xE0,0x00, \
  0x07,0xDF,0xC3,0xFB,0xF7,0x9F,0xFF,0x1E,0xFF,0xF9,0x03,0xF8,0x60,0x7F,0xFF,0xC0, \
  0x07,0xDF,0xC7,0xB9,0xFF,0xFF,0xF4,0x7F,0xFF,0xF8,0x07,0xB8,0xE0,0x67,0xFE,0x80, \
  0x0F,0xDF,0x87,0x31,0xFF,0x7E,0x00,0xFF,0xFC,0xF8,0x07,0x38,0xE0,0xF7,0xC0,0x00, \
  0x0F,0xFF,0x8F,0x7F,0xFE,0x1E,0x00,0xFF,0xF0,0xF0,0x0F,0x3F,0xE1,0xF3,0x80,0x00, \
  0x0E,0xEF,0x8F,0xFD,0xFE,0x1E,0x00,0xFF,0x01,0xF0,0x0F,0xFF,0xE3,0xE7,0x80,0x00, \
  0x0E,0x0F,0x1F,0xF9,0xFC,0x3C,0x03,0xFF,0x81,0xF0,0x3F,0xF9,0xC7,0xE7,0x80,0x70, \
  0x1E,0x1F,0x1F,0x71,0xFC,0x3C,0x7F,0xBF,0xE1,0xF0,0xEF,0x73,0xCF,0xC7,0x0F,0xF0, \
  0x1C,0x1F,0x1C,0x71,0xF8,0x3D,0xFF,0xBF,0xF9,0xF3,0xFC,0x73,0xDF,0x0F,0x7F,0xE0, \
  0x1C,0x1E,0x38,0x60,0xF0,0x3F,0xFF,0x3F,0xFD,0xFF,0xFC,0x63,0xFE,0x0F,0xFF,0xE0, \
  0x1C,0x1E,0x38,0xE0,0xE0,0x3F,0xFC,0x38,0xFF,0xFF,0xB8,0xE7,0xFC,0x07,0xFF,0x80, \
  0x1C,0x1E,0x30,0xE0,0xC0,0x3F,0xE0,0x38,0x0F,0x7F,0x30,0xEF,0xC0,0x07,0xF8,0x00, \
  0x00,0x3E,0x20,0xE0,0x00,0x1E,0x00,0x00,0x00,0x0E,0x20,0xE0,0x00,0x03,0xC0,0x00, \
  0x00,0x3C,0x00,0xC0,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0xC0,0x00,0x00,0x00,0x00, \
  0x00,0x3C,0x01,0xC0,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0xC0,0x00,0x00,0x00,0x00, \
  0x00,0x7C,0x01,0x40,0x00,0x00,0x00,0x00,0x00,0x00,0x01,0x40,0x00,0x00,0x00,0x00, \
  0x00,0x7C,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x7C,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x78,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x78,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0xF0,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0xC0,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x80,0x00,0x00,0x0F,0x3C,0x78,0xF8,0x31,0xF0,0x0E,0x78,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x10,0x32,0x04,0x08,0x50,0x10,0x10,0x40,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x10,0x22,0x05,0x08,0x52,0x10,0x10,0x40,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x10,0x22,0x09,0x08,0x92,0x10,0x10,0x60,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x13,0x36,0x31,0x08,0x92,0x13,0x90,0x60,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x11,0x30,0x41,0x09,0xFA,0x10,0x10,0x40,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x11,0x20,0x41,0x88,0x13,0x10,0x10,0x40,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x1F,0x20,0x7C,0xF8,0x11,0xF0,0x1E,0x78,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00 

#endif